File: eden/fs/service/Systemd.cpp
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */
#include "eden/fs/service/Systemd.h"
#include <folly/String.h>
#include <folly/logging/xlog.h>
#include <gflags/gflags.h>
#include "eden/fs/eden-config.h"

#if EDEN_HAVE_SYSTEMD
#include <systemd/sd-daemon.h> // @manual
#endif

namespace facebook {
namespace eden {

#if EDEN_HAVE_SYSTEMD
DEFINE_bool(
    experimentalSystemd,
    false,
    "EXPERIMENTAL: Run edenfs as if systemd controls its lifecycle");

void Systemd::notifyReady() {
  // TODO(strager): Move READY=1 into a systemd-specific StartupLogger.
  auto rc = sd_notify(/*unset_environment=*/false, "READY=1");
  if (rc < 0) {
    XLOG(ERR) << "sd_notify READY=1 failed: " << folly::errnoStr(-rc);
  } else if (rc == 0) {
    XLOG(WARN)
        << "sd_notify READY=1 failed: $NOTIFY_SOCKET is unset. edenfs was probably not started by systemd.";
  }
}
#endif

} // namespace eden
} // namespace facebook

File: eden/fs/cli_rs/edenfs-commands/src/debug/systemd.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

//! edenfsctl debug systemd

use anyhow::Context as _;
use anyhow::Result;
use async_trait::async_trait;
use clap::Parser;

use crate::ExitCode;
use crate::get_edenfs_instance;

/// Print the systemd unit name for this EdenFS instance.
#[derive(Parser, Debug)]
#[clap(about = "Print the systemd unit name for this EdenFS instance")]
pub struct SystemdCmd {}

#[async_trait]
impl crate::Subcommand for SystemdCmd {
    async fn run(&self) -> Result<ExitCode> {
        let instance = get_edenfs_instance();
        let unit = edenfs_client::daemon::get_systemd_unit(instance)
            .context("Failed to get systemd unit name")?;
        println!("{unit}");
        Ok(0)
    }
}
